flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/codegen/calls/ScalarOperators.scala

```diff
@@ -122,14 +125,14 @@ object ScalarOperators {
       right: GeneratedExpression)
     : GeneratedExpression = {
     generateOperatorIfNotNull(nullCheck, BOOLEAN_TYPE_INFO, left, right) {
-      if (isString(left) && isString(right)) {
+      if (isReference(left) && isReference(right)) {
         (leftTerm, rightTerm) => s"$leftTerm.compareTo($rightTerm) $operator 0"
       }
-      else if (isNumeric(left) && isNumeric(right)) {
+      else if (isNumeric(left.resultType) && isNumeric(right.resultType)) {
```

Comment:
If `left` or `right` is a `BigDecimal` invalid code is generated. 

> Caused by: org.codehaus.commons.compiler.CompileException: Line 64, Column 0: Cannot compare types "int" and "java.math.BigDecimal"

This happens if you replace the query in `FilterITCase.testFilterOnInteger()` by `SELECT * FROM MyTable WHERE a < 4.25`. `MyTable.a` is an `INTEGER` column and not auto-casted to `DECIMAL`. Auto-casting seems to work for expressions in the `SELECT` clause.